sys/src/9/ufs/ufs_mountpoint.c

```diff
@@ -0,0 +1,184 @@
+/*
+ * This file is part of the Harvey operating system.  It is subject to the
+ * license terms of the GNU GPL v2 in LICENSE.gpl found in the top-level
+ * directory of this distribution and at http://www.gnu.org/licenses/gpl-2.0.txt
+ *
+ * No part of Harvey operating system, including this file, may be copied,
+ * modified, propagated, or distributed except according to the terms
+ * contained in the LICENSE.gpl file.
+ */
+
+
+#include "u.h"
+#include "../../port/lib.h"
+#include "mem.h"
+#include "dat.h"
+#include "fns.h"
+
+#include "freebsd_util.h"
+#include "ufs_harvey.h"
+#include "ufs_ext.h"
+
+#include "ufs/quota.h"
+#include "ufs/inode.h"
+
+
+const static int VnodeFreelistBatchSize = 1000;
+
+static vnode*
+alloc_freelist()
+{
+        vnode *head = nil;
+        vnode *curr = nil;
+
+        for (int i = 0; i < VnodeFreelistBatchSize; i++) {
+                vnode *vn = mallocz(sizeof(vnode), 1);
+                if (vn == nil) {
+                       break;
+                }
+
+                if (head == nil) {
+                        head = vn;
+                }
+
+                vn->prev = curr;
+                if (curr != nil) {
+                        curr->next = vn;
+                }
+
+                curr = vn;
+        }
+        return head;
+}
+
+MountPoint*
+newufsmount(Chan *c, int id)
+{
+        MountPoint *mp = mallocz(sizeof(MountPoint), 1);
+        mp->chan = c;
+        mp->free_vnodes = nil;
+        mp->id = id;
+        return mp;
+}
+
+void
+releaseufsmount(MountPoint *mp)
+{
+        // No need to unlock later, since we're freeing mp
+        qlock(&mp->vnodes_lock);
+
+        // TODO HARVEY What if there are referenced vnodes?
```

Comment:
you'd probably have a ref in the mountpoint and sleep on it until it went to zero maybe? Not sure.